### PR TITLE
Testing capability on openshift

### LINKED ISSUE #1841: Integration tests for OpenShift
Given OpenShift's differences, particularly around security settings, we need integration tests to run as part of the CI.

### PR DESCRIPTION
**Pull request by petr-balogh**
This commit adding the required changes which were needed for running the
smoke tests on OCP 4 environment (3 hosts / 3 masters with 1 device
added on each of worker node).

Signed-off-by: Petr Balogh <pbalogh@redhat.com>

**Description of your changes:**
After discussion with @travisn I am pushing this PR to get some feedback what needs to be done in order to be able to run smoke tests on openshift environment.

I created new function to get proper CRs for rook operator openshift and have condition to use proper function depending on passed ROOK_TEST_OPENSHIFT environment variable.

If env variable is exported with export `ROOK_TEST_OPENSHIFT=true` then the tests are executed with proper changes to be able to run on openshift environment.

We still have some issues with CSI tests, issues opened by Travis here: #3517 and #3518 , currently this is the status of running smoke tests:

> Note: this is with disabled test: TestBlockCSI_SmokeTest as Travis suggested cause of issues mentioned above
```bash
--- FAIL: TestCephSmokeSuite (1775.59s)
    --- PASS: TestCephSmokeSuite/TestBlockStorage_SmokeTest (168.04s)
    --- PASS: TestCephSmokeSuite/TestFileStorageMountUser_SmokeTest (83.41s)
    --- FAIL: TestCephSmokeSuite/TestFileStorage_SmokeTest (991.69s)
        nfs.go:50:
            	Error Trace:	nfs.go:50
            	            				ceph_base_file_test.go:56
            	            				ceph_base_file_test.go:47
            	            				ceph_smoke_test.go:96
            	Error:      	Expected nil, but got: &errors.errorString{s:"Giving up waiting for pod with label ceph_nfs=my-nfs in namespace smoke-ns to be running"}
            	Test:       	TestCephSmokeSuite/TestFileStorage_SmokeTest
        nfs.go:52:
            	Error Trace:	nfs.go:52
            	            				ceph_base_file_test.go:56
            	            				ceph_base_file_test.go:47
            	            				ceph_smoke_test.go:96
            	Error:      	Should be true
            	Test:       	TestCephSmokeSuite/TestFileStorage_SmokeTest
            	Messages:   	Make sure all nfs daemon pods are in Running state
    --- PASS: TestCephSmokeSuite/TestMonFailover (42.05s)
    --- PASS: TestCephSmokeSuite/TestObjectStorage_SmokeTest (113.51s)
    --- PASS: TestCephSmokeSuite/TestPoolResize (21.31s)
    --- PASS: TestCephSmokeSuite/TestRookClusterInstallation_SmokeTest (1.78s)
FAIL
FAIL	github.com/rook/rook/tests/integration	1775.955s
```

**Which issue is resolved by this Pull Request:**
Resolves #1841

**Checklist:**

- [x] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [x] Documentation has been updated, if necessary.
- [x] Unit tests have been added, if necessary.
- [x] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.
- [ ] Comments have been added or updated based on the standards set in [CONTRIBUTING.md](https://github.com/rook/rook/blob/master/CONTRIBUTING.md#comments)
- [ ] Add the flag for skipping the CI if this PR does not require a build. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for more details.

[test ceph]
[test full]

### CODE REVIEW

File: tests/framework/installer/ceph_manifests.go [hunk 1]
```diff
@@ -565,6 +572,921 @@ subjects:
...
+        configMap:
+          name: csi-rbd-config
+      - name: csi-cephfs-config
+        configMap:
+          name: csi-cephfs-config
+`
+}
+
+// GetRookOperatorForOpenshift returns rook Operator manifest for openshift
+func getRookOperatorForOpenshiftYaml(namespace string, m *CephManifestsMaster) string {
```
Comment (petr-balogh) [inline tests/framework/installer/ceph_manifests.go]: @travisn I remember from you comment in my fork [here](https://github.com/petr-balogh/rook/commit/163216fedc75c90a1c4b74e68869f377403f008a#r34344737) that you would like to have separate function and have the yaml defined there for openshift for better readability than compose this yaml because of some parts are different.

In this approach I see a lot of duplicity and it's probably not good for maintenance as ppl will need to add for both places.  Do you have in mind different solution you prefer here?

File: tests/framework/installer/installer.go [hunk 2]
```diff
@@ -85,10 +86,19 @@ func init() {
...
 	}
+	if RookTestOpenshift {
+		baseTestDir = "/var/lib/rook"
+		fmt.Sprintf("baseTestDir set to %s", baseTestDir)
+	}
 }
 
 func SystemNamespace(namespace string) string {
-	return fmt.Sprintf("%s-system", namespace)
+	if RookTestOpenshift {
```
Comment (petr-balogh) [inline tests/framework/installer/installer.go]: I remember it didn't work for me when it was in two namespaces even if I changed system:serviceaccount: NAMESPACE in SCC definition.  Do you know what can cause the issues on openshift? Or do you think of any issue if we will use just one NS when testing on OCP @travisn ?

File: tests/framework/installer/ceph_manifests.go (hunk 1 above)
Comment (travisn) [reply tests/framework/installer/ceph_manifests.go]: Agreed we don't want to duplicate all of this. The main differences are the two SCCs and an env var in the rook-ceph-operator deployment. Sounds great if we can just add conditionals to add those for openshift.

File: tests/framework/installer/installer.go (hunk 2 above)
Comment (travisn) [reply tests/framework/installer/installer.go]: - One issue with OpenShift can be that working across namespaces might be disabled by default. #2723 has the documentation for working with that.
- Working with a single namespace is a good place to start. Working with multiple namespaces just ensures that scenario is working. Ideally we would have a mix of tests using both options.

File: tests/framework/installer/ceph_manifests.go (hunk 1 above)
Comment (petr-balogh) [reply tests/framework/installer/ceph_manifests.go]: @travisn I changed the logic here to do not duplicate the code, is it OK like this?

**Comment by petr-balogh** [comment]
@travisn on latest nightly openshift 4.2 I see Smoke Suite passed OK:
--- PASS: TestCephSmokeSuite (753.27s)
    --- PASS: TestCephSmokeSuite/TestBlockCSI_SmokeTest (0.67s)
    --- PASS: TestCephSmokeSuite/TestBlockStorage_SmokeTest (174.61s)
    --- PASS: TestCephSmokeSuite/TestFileStorageMountUser_SmokeTest (68.90s)
    --- PASS: TestCephSmokeSuite/TestFileStorage_SmokeTest (74.44s)
    --- PASS: TestCephSmokeSuite/TestMonFailover (42.32s)
    --- PASS: TestCephSmokeSuite/TestObjectStorage_SmokeTest (101.20s)
    --- PASS: TestCephSmokeSuite/TestPoolResize (14.85s)
    --- PASS: TestCephSmokeSuite/TestRookClusterInstallation_SmokeTest (2.33s)
PASS
--- PASS: TestCephBlockCreateSuite (382.16s)
    --- PASS: TestCephBlockCreateSuite/TestBlockStorageMountUnMountForStatefulSets (86.44s)
    --- PASS: TestCephBlockCreateSuite/TestCreatePVCWhenNoStorageClassExists (7.12s)
    --- PASS: TestCephBlockCreateSuite/TestCreatingPVCWithVariousAccessModes (55.02s)
PASS

--- FAIL: TestCephUpgradeSuite (407.38s)
--- FAIL: TestCephBlockMountUnMountSuite (461.12s)
    --- FAIL: TestCephBlockMountUnMountSuite/TestBlockStorageMountUnMountForDifferentAccessModes (218.55s)
        ceph_zblock_mount_unmount_test.go:267:
            	Error Trace:	ceph_zblock_mount_unmount_test.go:267
            	Error:      	Expected nil, but got: &errors.errorString{s:"failed to read file bsFile1 from pod rwo-block-ro-one. failed to read file bsFile1 from pod rwo-block-ro-one. Failed to run: oc [exec rwo-block-ro-one -- cat /tmp/testrook/bsFile1] : Failed to complete 'oc': exit status 1. "}
            	Test:       	TestCephBlockMountUnMountSuite/TestBlockStorageMountUnMountForDifferentAccessModes
        ceph_zblock_mount_unmount_test.go:268:
            	Error Trace:	ceph_zblock_mount_unmount_test.go:268
            	Error:      	Expected nil, but got: &errors.errorString{s:"failed to read file bsFile1 from pod rwo-block-ro-two. failed to read file bsFile1 from pod rwo-block-ro-two. Failed to run: oc [exec rwo-block-ro-two -- cat /tmp/testrook/bsFile1] : Failed to complete 'oc': exit status 1. "}
            	Test:       	TestCephBlockMountUnMountSuite/TestBlockStorageMountUnMountForDifferentAccessModes
        ceph_zblock_mount_unmount_test.go:271:
            	Error Trace:	ceph_zblock_mount_unmount_test.go:271
            	Error:      	Expected nil, but got: &errors.errorString{s:"failed to read file bsFile1 from pod rwx-block-ro-one. failed to read file bsFile1 from pod rwx-block-ro-one. Failed to run: oc [exec rwx-block-ro-one -- cat /tmp/testrook/bsFile1] : Failed to complete 'oc': exit status 1. "}
            	Test:       	TestCephBlockMountUnMountSuite/TestBlockStorageMountUnMountForDifferentAccessModes
        ceph_zblock_mount_unmount_test.go:272:
            	Error Trace:	ceph_zblock_mount_unmount_test.go:272
            	Error:      	Expected nil, but got: &errors.errorString{s:"failed to read file bsFile1 from pod rwx-block-ro-two. failed to read file bsFile1 from pod rwx-block-ro-two. Failed to run: oc [exec rwx-block-ro-two -- cat /tmp/testrook/bsFile1] : Failed to complete 'oc': exit status 1. "}
            	Test:       	TestCephBlockMountUnMountSuite/TestBlockStorageMountUnMountForDifferentAccessModes
FAIL
FAIL	github.com/rook/rook/tests/integration	461.219s


So on openshift env we have at least:
TestCephSmokeSuite and TestCephBlockCreateSuite passing.

So I suggest to cover the rest in different PR if possible?

I see that now the CI is also looks much better but still some of test didn't pass. @travisn I remember you said that you sometime have some random failures, is this the case?

**Comment by petr-balogh** [comment]
@travisn for now I skipped tests on openshift env which had issues. We have passing 2 suites:
TestCephSmokeSuite - Status PASS - but we have 1 test disabled cause of issues you pointed in mail thread and mentioned in comment in [commit](https://github.com/rook/rook/pull/3523/commits/02c1de135a2f1e0db87e5089de2b1a3c22918a59), referring them also here #3517, #3151 
TestCephBlockCreateSuite - Status PASS

I am suggesting to merge this one and then the rest of tests can be solved in separate PRs suite by suite WDYT @travisn.  Each of skip is done in separate commit, so those can be reverted in the following PRs.

Thanks

**Comment by petr-balogh** [comment]
I see that CI is passing here and I was asked to add some documentation. So I did it in different PR as it maybe will require some other changes after review and I don't want to block this PR cause of this. Doc PR is opened here #3637

File: tests/framework/installer/installer.go [hunk 3]
```diff
@@ -85,10 +87,19 @@ func init() {
 		createBaseTestDir = false
 		baseTestDir = "/data"
 	}
+	if RookTestOpenshift {
+		baseTestDir = "/var/lib/rook"
```
Comment (travisn) [inline tests/framework/installer/installer.go]: Why does this need to be `/var/lib/rook` on openshift? Can we make this an environment variable instead of hard coding for openshift?

File: tests/framework/installer/installer.go [hunk 4]
```diff
@@ -85,10 +87,19 @@ func init() {
...
+		fmt.Sprintf("baseTestDir set to %s", baseTestDir)
+	}
 }
 
 func SystemNamespace(namespace string) string {
-	return fmt.Sprintf("%s-system", namespace)
+	if RookTestOpenshift {
+		fmt.Sprintf("For openshift execution used system namespace: %s", namespace)
```
Comment (travisn) [inline tests/framework/installer/installer.go]: instead of `fmt.Sprintf()`, please use `logger.Infof()`

File: tests/framework/utils/k8s_helper.go [hunk 5]
```diff
@@ -63,6 +61,16 @@ const (
 	hostnameTestPrefix = "test-prefix-this-is-a-very-long-hostname-"
 )
 
+// GetKubectlCmd returns kubectl or oc if env var ROOK_TEST_OPENSHIFT is
+// set to true
+func GetKubectlCmd() string {
+	var kubectl_cmd = "kubectl"
```
Comment (travisn) [inline tests/framework/utils/k8s_helper.go]: nit: a more common golang var init would be:
```
kubectl_cmd := "kubectl"
```

File: tests/framework/utils/k8s_helper.go [hunk 6]
```diff
@@ -86,7 +94,23 @@ func CreateK8sHelper(t func() *testing.T) (*K8sHelper, error) {
...
+	RetryLoop = getRetryLoopNumber()
+	// RookTestOpenshift variable which define testing on openshift
+	RookTestOpenshift = os.Getenv("ROOK_TEST_OPENSHIFT") == "true"
+)
+
+func getRetryLoopNumber() int {
+	// For openshift execution we need to increase retry loop as deployment
+	// takes more time
+	if RookTestOpenshift {
+		return 60
```
Comment (travisn) [inline tests/framework/utils/k8s_helper.go]: Instead of a single openshift environment variable, how about a different env variable for settings like this? The retry loop doesn't depend on openshift, but really depends on the speed of the cluster.

File: tests/integration/ceph_helm_test.go [hunk 7]
```diff
@@ -46,7 +46,7 @@ var (
 // - Create the object store via the CRD
 // ***************************************************
 func TestCephHelmSuite(t *testing.T) {
-	if installer.SkipTestSuite(installer.CephTestSuite) {
+	if installer.SkipTestSuite(installer.CephTestSuite) || RookTestOpenshift {
```
Comment (travisn) [inline tests/integration/ceph_helm_test.go]: This is another place where a separate environment variable would be preferred instead of `RookTestOpenshift`. It could be something like this:
```
if installer.SkipTestSuite(installer.CephTestSuite) || os.Getenv("ROOK_SKIP_CEPH_HELM_SUITE") == "true" {
  t.Skip()
}
```

Similarly for other tests that are being skipped for openshift. This flag could also be used in the future for some other test environment besides openshift.

File: tests/integration/ceph_smoke_test.go [hunk 8]
```diff
@@ -87,6 +93,10 @@ func (suite *SmokeSuite) TearDownSuite() {
 }
 
 func (suite *SmokeSuite) TestBlockCSI_SmokeTest() {
+	if RookTestOpenshift {
+		logger.Info("Skipping this test on OpenShift ENV because it's blocked by github issues: 3517 and 3520")
```
Comment (travisn) [inline tests/integration/ceph_smoke_test.go]: #3517 and #3520 are closed now (merged with #3373). Can we enable this test agin?

File: tests/framework/installer/installer.go [hunk 9]
```diff
@@ -46,6 +46,8 @@ var (
 	baseTestDir       string
 	forceUseDevices   = false
 	createBaseTestDir = true
+	// RookTestOpenshift variable which define testing on openshift
+	RookTestOpenshift = os.Getenv("ROOK_TEST_OPENSHIFT") == `true`
```
Comment (travisn) [inline tests/framework/installer/installer.go]: Instead of adding RookTestOpenshift var here, how about adding it to the EnvironmentManifest struct [here](https://github.com/rook/rook/blob/master/tests/framework/installer/environment.go)? Then we can unify the new env vars with the existing ones.

File: tests/framework/installer/installer.go (hunk 3 above)
Comment (petr-balogh) [reply tests/framework/installer/installer.go]: @travisn this I saw is set here: 
https://github.com/rook/rook/blob/82702f5cc5d6150fbcc3990045a75379338b1d96/cluster/examples/kubernetes/ceph/cluster.yaml#L28-L31
like this as well and as I remember correctly I was failing on openshift when it was set to /data . So I am using what is set also in mentioned example as I see /data is mentioned for minikube.

Comment (travisn) [reply tests/framework/installer/installer.go]: @petr-balogh What about a new env variable? The test directory could be different for various OpenShift or Kubernetes environments, so it doesn't seem like we should tie it to the RookTestOpenshift variable.  
```
  if RookBaseTestDir != "" {
     baseTestDir = RookBaseTestDir
     fmt.Sprintf("baseTestDir set to %s", baseTestDir)
  } 
```

**Comment by petr-balogh** [comment]
Hello @travisn , I didn't have time to work on changes here last week and was out of the office on Friday. I've tried to address your comments from last review, so please check and comment if something is missing.  I will need to retest again cause seems that it's still doesn't work 100 % well on openshift. So need to investigate and spend some time on debugging tomorrow. And also want to see how it will pass the CI here or if I broke something else by changes.

File: tests/framework/environment/environment.go [hunk 10]
```diff
@@ -50,4 +57,12 @@ func init() {
...
 	flag.BoolVar(&Env.EnableChaos, "enable_chaos", false, "used to determine if random pods in a namespace are to be killed during load test.")
 	flag.StringVar(&Env.Logs, "logs", "", "Gather rook logs, eg - all")
+	flag.BoolVar(&Env.RookTestOpenshift, "rook_test_openshift", false, "used to determine if running agains openshift ENV.")
+	flag.IntVar(&Env.RetryLoopNumber, "retry_loop", 40, "number of attempts for retry in the loop.")
+	flag.StringVar(&Env.RookBaseTestDir, "rook_base_test_dir", "", "base test dir of rook.")
+	flag.BoolVar(&Env.SkipCephHelmSuite, "skip_ceph_helm_suite", false, "skip CephHelmSuite")
+	flag.BoolVar(&Env.SkipCephMultiClusterDeploySuite, "skip_ceph_multi_cluster_deploy_suite", false, "skip CephMultiClusterDeploySuite")
+	flag.BoolVar(&Env.SkipCephUpgradeSuite, "skip_ceph_upgrade_suite", false, "skip CephUpgradeSuite")
+	flag.BoolVar(&Env.SkipCephBlockMountUnMountSuite, "skip_ceph_block_mount_unmount_suite", false, "skip CephBlockMountUnMountSuite")
+	flag.Parse()
```
Comment (petr-balogh) [inline tests/framework/environment/environment.go]: @travisn I was kind of surprised that this `flag.Parse()` wasn't called before cause without this call I couldn't get value of any passed arguments

File: tests/framework/installer/ceph_manifests.go [hunk 11]
```diff
@@ -476,9 +477,110 @@ spec:
...
+  - driver: "ceph.rook.io/rook"
+  - driver: "ceph.rook.io/rook-ceph"
+volumes: ['*']
+users:
+  # A user needs to be added for each rook service account.
+  # This assumes running in the default sample "rook-ceph" namespace.
+  # If other namespaces or service accounts are configured, they need to be updated here.
+  - system:serviceaccount:` + namespace + `:rook-csi-rbd-plugin-sa
+  - system:serviceaccount:` + namespace + `:rook-csi-rbd-provisioner-sa
+  - system:serviceaccount:` + namespace + `:rook-csi-rbd-attacher-sa
```
Comment (Madhu-1) [inline tests/framework/installer/ceph_manifests.go]: attacher-sa is not required,  this PR removes it https://github.com/rook/rook/pull/3768

**Comment by obnoxxx** [comment]
@petr-balogh if you don't have time to pursue this PR to completion, maybe @Madhu-1 can help out!

**Comment by petr-balogh** [comment]
> @petr-balogh if you don't have time to pursue this PR to completion, maybe @Madhu-1 can help out!

I had discussion with @RazTamir and if this one be taken by someone from engineering I will be happy, I was working on another important stuff in QE like deployment via operator upstream and downstream. I started also working on automation for upgrade so not sure if I will have time to process quickly with this as this is really time consuming cause of investigating of all the issues with lack of full understanding of testing framework and all the tests.

After last changes of implement the comments from @travisn I saw that the tests stopped passing like before but will need more investigation. If someone else will be assigned to this one @obnoxxx I can set the mtg with him/her and share my knowledge of what was done and so on. 

For sure this will require new rebase and align with latest comment and investigate and fix the rest issues at least for the smoke suite for now.

**Comment by Madhu-1** [comment]
@petr-balogh @obnoxxx I will  address all review  comments and test this one on openshift

**Comment by Madhu-1** [comment]
> > @petr-balogh if you don't have time to pursue this PR to completion, maybe @Madhu-1 can help out!
> 
> I had discussion with @RazTamir and if this one be taken by someone from engineering I will be happy, I was working on another important stuff in QE like deployment via operator upstream and downstream. I started also working on automation for upgrade so not sure if I will have time to process quickly with this as this is really time consuming cause of investigating of all the issues with lack of full understanding of testing framework and all the tests.
> 
> After last changes of implement the comments from @travisn I saw that the tests stopped passing like before but will need more investigation. If someone else will be assigned to this one @obnoxxx I can set the mtg with him/her and share my knowledge of what was done and so on.
> 
> For sure this will require new rebase and align with latest comment and investigate and fix the rest issues at least for the smoke suite for now.

having a meeting really helps as I need some input on setting up the environment to test it

**Comment by petr-balogh** [comment]
@Madhu-1 I sent you mtg invitation for tomorrow morning. 9:30am my time, if the time is not ok for you just reply to the invitation which time you suggest and I can update the time if fit tom me. Thanks

File: tests/framework/environment/environment.go (hunk 10 above)
Comment (Madhu-1) [reply tests/framework/environment/environment.go]: as this package name was `installer` earlier it was called here https://github.com/rook/rook/blob/ba74cc9361bbccd2d9f94411b826538967bf11d3/tests/framework/installer/ceph_installer.go#L579

Comment (petr-balogh) [reply tests/framework/environment/environment.go]: @BlaineEXE The intention is to run all tests starting with: TestCeph
Before the rebase and align with Travis comments we had passing those two test suites:
- TestCephSmokeSuite
- TestCephBlockCreateSuite

I had the conditions different way and run just those tests which were already unblocked if openshift testing parameter passed. But then @travisn commented to use this approach: https://github.com/rook/rook/pull/3523#discussion_r315422213 . So I did for each of those tests suites which do not pass the parameter to have possibility to skip them as was requested by Travis.

**Comment by leseb** [comment]
> @travisn @leseb my testing is struck with below logs
> 
> ```
> 2019-09-12 15:39:27.830642 I | testutil: Running oc [create -f -]
> objectbucketclaim.objectbucket.io/smoke-delete-bucket created
> 2019-09-12 15:39:29.038902 I | exec: Running command: oc get obc smoke-delete-bucket
> 2019-09-12 15:39:30.322835 I | clients: obc smoke-delete-bucket created
> 2019-09-12 15:39:30.322936 I | exec: Running command: oc get secret smoke-delete-bucket
> 2019-09-12 15:39:31.480999 I | clients: secret smoke-delete-bucket created
> 2019-09-12 15:39:31.481029 I | exec: Running command: oc get configmap smoke-delete-bucket
> 2019-09-12 15:39:32.664148 I | clients: configmap smoke-delete-bucket created
> 2019-09-12 15:39:32.665196 I | clients: smoke-delete-bucket resources [obc secret configmap] all created
> 2019-09-12 15:39:32.665467 I | integrationTest: Check if bucket was created
> 2019-09-12 15:39:32.665668 I | exec: Running command: kubectl -it exec rook-ceph-tools -n smoke-ns -- radosgw-admin bucket stats --bucket smokebkt --rgw-realm=teststore --rgw-zonegroup=teststore
> 2019-09-12 15:39:36.409960 I | exec: Running command: kubectl -it exec rook-ceph-tools -n smoke-ns -- radosgw-admin metadata get bucket:smokebkt --rgw-realm=teststore --rgw-zonegroup=teststore
> 2019-09-12 15:39:39.669799 I | integrationTest: OBC, Secret and ConfigMap created
> 2019-09-12 15:39:39.669886 I | integrationTest: Step 4 : Create s3 client
> 2019-09-12 15:39:39.670549 I | exec: Running command: oc get svc -n smoke-ns -l rgw=teststore -o jsonpath={.items[*].spec.clusterIP}
> 2019-09-12 15:39:40.915350 I | exec: Running command: oc get secret smoke-delete-bucket -o jsonpath={@.data.AWS_ACCESS_KEY_ID}
> 2019-09-12 15:39:42.169091 I | exec: Running command: oc get secret smoke-delete-bucket -o jsonpath={@.data.AWS_SECRET_ACCESS_KEY}
> 2019-09-12 15:39:43.383574 I | integrationTest: endpoint (172.30.245.56:80) Accesskey (69XGCU309O50EV2DB3EP) secret (cBAXTTg4K32Ntprh42OL96vOZPhOT6dM1GmI3eQv)
> 2019-09-12 15:39:43.383595 I | integrationTest: Step 5 : Put Object on bucket
> ```

@jeffvance any idea?

**Comment by Madhu-1** [comment]
@thotz @copejon any idea without the above behavior? what logs I need to check and how to debug this.

**Comment by jeffvance** [comment]
@Madhu-1 Jon is on PTO today. I don't see an error where you're stuck above. Is there a way for you to run the test with a higher log level (>=2)?

**Comment by Madhu-1** [comment]
@jeffvance Thanks for the reply tests are running with `-v` flag, this is what I can see it.

@travisn is it way to set log level in testing?

@jeffvance the  `Step 5 : Put Object on bucket` is struck for more than 20 min in my testing (3 times). if I want to debug it can you tell me which container log I can check for this

**Comment by jeffvance** [comment]
@Madhu-1 I run provisioners with `-v=2 -alsologtostderr` but I don't know how you set these flags in your tests

**Comment by thotz** [comment]
@Madhu-1 : I have never run the test in open shift env may it is related to that.  On minikube tests were passing last week. As @jeffvance suggested please get the container logs for the rook operator

**Comment by Madhu-1** [comment]
@thotz attaching [operator.log](https://github.com/rook/rook/files/3616303/operator.log)

**Comment by jeffvance** [comment]
From log above:
`2019-09-16 11:57:18.283784 I | op-bucket-prov: successfully created bucket "smokebkt"`
shows that the bucket was created by the rook-ceph bucket operator.
And:
```
I0916 11:57:18.601508       6 resourcehandlers.go:139]  "level"=0 "msg"="creating Secret" "key"="default/smoke-delete-bucket" "name"="default/smoke-delete-bucket"
I0916 11:57:18.669240       6 resourcehandlers.go:162]  "level"=0 "msg"="creating ConfigMap" "key"="default/smoke-delete-bucket" "name"="default/smoke-delete-bucket"
I0916 11:57:18.673339       6 helpers.go:78]  "level"=0 "msg"="getting claim for key" "key"="default/smoke-delete-bucket" 
I0916 11:57:18.675799       6 resourcehandlers.go:119]  "level"=0 "msg"="creating ObjectBucket" "key"="default/smoke-delete-bucket" "name"="obc-default-smoke-delete-bucket"
I0916 11:57:18.681154       6 resourcehandlers.go:273]  "level"=0 "msg"="updating status:" "key"="default/smoke-delete-bucket" "new status"="bound" "ob"="obc-default-smoke-delete-bucket" "old status"=""
I0916 11:57:18.685809       6 resourcehandlers.go:252]  "level"=0 "msg"="updating" "key"="default/smoke-delete-bucket" "obc"="default/smoke-delete-bucket"
I0916 11:57:18.690222       6 resourcehandlers.go:261]  "level"=0 "msg"="updating status:" "key"="default/smoke-delete-bucket" "new status"="bound" "obc"="default/smoke-delete-bucket" "old status"="pending"
I0916 11:57:18.695575       6 controller.go:333]  "level"=0 "msg"="provisioning succeeded" "key"="default/smoke-delete-bucket" 
```
shows that the secret, CM and OB were created by the bucket lib.
But this log line at the end I haven't seen before:
`W0916 12:07:05.354356       6 reflector.go:289] github.com/rook/rook/pkg/operator/ceph/cluster/controller.go:165: watch of *v1.ConfigMap ended with: too old resource version: 70771 (74419)`

**Comment by thotz** [comment]
@Madhu-1 , I tried to set up minishift but failed on my machine. As Jeff mentioned OBC and everything got created if possible can u try to do the creating objects manually post obc creation. Or else can u please provide me with a set up in where I can  see what went wrong

**Comment by Madhu-1** [comment]
> @Madhu-1 , I tried to set up minishift but failed on my machine. As Jeff mentioned OBC and everything got created if possible can u try to do the creating objects manually post obc creation. Or else can u please provide me with a set up in where I can see what went wrong

sending you the kubeconfig in mail

**Comment by copejon** [comment]
@jeffvance 

TL;DR  - `too old resource version` is a well known issue with k8s with a huge number of issues going back to k8s v1.1.0.  It's probably a red herring.

The `too old resource version` is apparently a known k8s issue with CRD watchers: [#55230](https://github.com/kubernetes/kubernetes/issues/55230).  It also existed, but was reportedly fixed, in k8s as far back as 2016: #[22830](https://github.com/kubernetes/kubernetes/issues/22830).  It's hard to discern what the actual cause is - one issues says it's a consequence of quick consecutive updates to a resource, causing the watcher to fall behind.  Another attributes it to the watcher cache being too small.   Yet another says that it's actually an error generated by the etcd server and mis-reported to the apiserver with a 500 http response, causing it to be passed on to the client.  One open PR seeks to fix this: [Increase watcher cache](https://github.com/kubernetes/kubernetes/issues/75677)

@Madhu-1 using the rook-ceph toolbox or operator pod, can you install s3cmd and attempt to put an object into the bucket manually?

Besides the rook-ceph-operator pod, are there any significant entries in the object pod?

**Comment by Madhu-1** [comment]
@jeffvance @copejon @travisn today I worked with @thotz to test are we able to access the object store from toolbox pod. we have installed `awscli` in the toolbox pod we are able to put data into the bucket with from the toolbox pod. 

**note**: the service created from accessing the store will be of type clusterIP

but in the E2E testing, we creating the object store client from  URL contains clusterIP and port because of this issue ( as we are using clusterIP:Port which cannot be accessed from the test machine ) we are not able to put data into the bucket?

**Comment by copejon** [comment]
@madhu-1 So from the CLI, you can use put data into the bucket.  But the tests cannot reach the service endpoint?  What is the specific error you're getting?  Looking back to the original tests, I don't see any object errors.

**Comment by Madhu-1** [comment]
> @Madhu-1 So from the CLI, you can use put data into the bucket. But the tests cannot reach the service endpoint? What is the specific error you're getting? Looking back to the original tests, I don't see any object errors.

@copejon tests are running from an external machine which is connecting to the openshift cluster deployed in aws. as this is `clusterIP` service which cannot be connected by external machine, I receive a connection timeout error. this I have found out while doing manual testing with awscli.

**Comment by copejon** [comment]
@Madhu-1 Ah! Okay, that makes sense. 

The case then is that the object store needs to be exposed to the outside world.  I'm not the guy to ask about networking, particularly in AWS, unfortunately.

**Comment by copejon** [comment]
Via a little testing in aws, the following seems to work: 

`kubectl expose pod -n rook-ceph rook-ceph-rgw-my-store-a-58675c6464-rzr6z --type=LoadBalancer --port=80`

You can then get the generated URL by:
`kubectl get svc -n rook-ceph`

and should see something like:
```
NAME                                                TYPE           CLUSTER-IP       EXTERNAL-IP                                                              PORT(S)             AGE
service/rook-ceph-rgw-my-store-a-58675c6464-rzr6z   LoadBalancer   100.71.190.214   a3e91aa09dbcb11e9827e0edfca08e4f-143351406.us-east-1.elb.amazonaws.com   80:31809/TCP        9m2s
```

The External-IP is an externally routable endpoint to the object store.

**Comment by Madhu-1** [comment]
> Via a little testing in aws, the following seems to work:
> 
> `kubectl expose pod -n rook-ceph rook-ceph-rgw-my-store-a-58675c6464-rzr6z --type=LoadBalancer --port=80`
> 
> You can then get the generated URL by:
> `kubectl get svc -n rook-ceph`
> 
> and should see something like:
> 
> ```
> NAME                                                TYPE           CLUSTER-IP       EXTERNAL-IP                                                              PORT(S)             AGE
> service/rook-ceph-rgw-my-store-a-58675c6464-rzr6z   LoadBalancer   100.71.190.214   a3e91aa09dbcb11e9827e0edfca08e4f-143351406.us-east-1.elb.amazonaws.com   80:31809/TCP        9m2s
> ```
> 
> The External-IP is an externally routable endpoint to the object store.

@copejon i was busy with other work i will test this out soon and make required changes to make it work with OCP

**Comment by jeffvance** [comment]
@petr-balogh looks like a rebase is needed

**Comment by petr-balogh** [comment]
> @petr-balogh looks like a rebase is needed

@jeffvance we agreed that this will be taken from here by engineering. @Madhu-1  did you do some progress here?  I do not see you pushed any changes to my branch this branch yet to update this PR.

**Comment by RazTamir** [comment]
Do we have any progress here?

**Comment by petr-balogh** [comment]
@Madhu-1  hello, did you do some progress here, I didn't touch it from time you started looking into it. Thanks for update.  @RazTamir I am not sure about progress since we decided to pass it to engineering.

**Comment by Madhu-1** [comment]
@petr-balogh @RazTamir sorry I cannot continue on this PR anymore because am busy with CSI 2.0 work and i conveyed same to @obnoxxx as well,

**Comment by travisn** [comment]
Had a chat with @bipuladh to take ownership of this now. Thanks!

**Comment by bipuladh** [comment]
@leseb Got busy with other tasks and completely forgot about it. If there is still time I can start work on it again.

**Comment by travisn** [comment]
@bipuladh Thanks it would be great if you could pick this up again. I'd suggest just waiting until #4853 is merged since it will likely cause a number of merge conflicts.

**Comment by bipuladh** [comment]
I see https://github.com/rook/rook/pull/4853 has merged will start work on it.

**Comment by leseb** [comment]
@bipuladh please rebase.

**Comment by leseb** [comment]
Closing this due to inactivity. Feel free to re-open. Thanks. I'd still love to have this at some point though.

**Comment by subhamkrai** [comment]
#6185
